Implemented conversion from networkx graph to latex graph.

LucaCappelletti94 opened the pull request:
Implemented conversion from networkx graph to latex updating my old implementation of [pyadigraph](https://github.com/LucaCappelletti94/pyadigraph), converting arbitrary graphs into [Adigraph pure tikz graphs](https://ctan.org/pkg/adigraph).

Example:
![Adigraph](https://github.com/LucaCappelletti94/pyadigraph/raw/master/example.png?raw=true)

dschult commented:
Wow --- This looks really nice.
I've been focusing on releasing v2.4 and will get more time soon.
Thanks!

LucaCappelletti94 commented:
Any updates on this pull request?

dschult commented:
Thanks for the gentle nudge!   
No updates except -- wow why haven't I gotten to this yet?

pep8speaks commented:
Hello @LucaCappelletti94! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`networkx/drawing/__init__.py`](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/__init__.py):

> [Line 7:1](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/__init__.py#L7): [F401](https://duckduckgo.com/?q=pep8%20F401) '.nx_latex.Adigraph' imported but unused

* In the file [`networkx/drawing/nx_latex.py`](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py):

> [Line 12:1](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L12): [F401](https://duckduckgo.com/?q=pep8%20F401) 'typing.Union' imported but unused
> [Line 147:17](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L147): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 148:17](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L148): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 150:17](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L150): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 151:17](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L151): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 153:17](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L153): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 154:17](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L154): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 162:22](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L162): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 175:19](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L175): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 225:13](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L225): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 246:13](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L246): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 338:89](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L338): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 88 characters)
> [Line 342:89](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L342): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 88 characters)
> [Line 348:13](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/nx_latex.py#L348): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line

* In the file [`networkx/drawing/tests/test_adigraph.py`](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/tests/test_adigraph.py):

> [Line 2:1](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/tests/test_adigraph.py#L2): [F401](https://duckduckgo.com/?q=pep8%20F401) 'random' imported but unused
> [Line 5:1](https://github.com/networkx/networkx/blob/8822cc1692d73b2ae377263e1c8c82bf0c724a3e/networkx/drawing/tests/test_adigraph.py#L5): [F401](https://duckduckgo.com/?q=pep8%20F401) 'nose.SkipTest' imported but unused

dschult commented:
I removed the conflicts with the master branch and reformatted docstrings and switched names from vertices/vertex to nodes/node.  Also format changes to match PEP8, and removed numpy requirements for the tests, as well as making the test class into a test function. I also made some cosmetic code changes.

Questions/concerns/suggestions:
- Should we make a public function like ```write_latex(G, path)``` to ease use?  
- There are other aspects of the public-facing API that raise questions... Does the basic code need to be inside a figure environment?  Should it use subfigures at all (that's for putting together many graphs... which presumably should be done by another program using this one to generate the latex code to draw just the one figure).
- Why does the code cut the x and y layout sizes in half?
- Could we remove the placeholder files and replace them with text strings in nx_latex.py? The files are small.  Is there another functionality you aim to include by allowing placeholder files?
- Could the test compare the strings generated rather than writing them to files and comparing those?  I think  this means, including the expected.tex file as a string and comparing that string to the string generated by ```str(A)```.  
- Will this code work for DiGraph and MultiGraph?
- Can we make some examples in the doc_string? Or in the examples gallery (directory examples/drawing)?

Thanks!

dschult commented:
Looks like the Appveyor tests were having trouble due to matching files with vs without ending linefeeds.  It might have also been differences in how files end their lines: with \n or \n\r. But I can't tell for sure (and I don't have access to a windows machine for local testing). 

The current test I have constructed writes a file and then reads and compares each line of the written test file and an expected output file. This might be better done using a string rather than writing/reading to files. But it seems to work OK and alleviates the difficulties with trailing end-of-line issues.

Any thoughts on better ways to test this?

dschult commented:
I can't get LaTeX to compile using the ```expected.tex``` file. It gives an error about too many "," tokens from xparse.  

Do I need to install anything, or use a particular flavor of pdflatex/latex/etc to get this to work?

dschult commented:
@LucaCappelletti94 can you verify that the ```expected.tex``` file that matches the output of ```A.save``` is working latex code? 
I'm trying to use ```pdflatex example.tex``` and it gives a syntax error about extra commas.

LucaCappelletti94 commented:
Hello @dschult and sorry for the late reply. I am checking it now.

LucaCappelletti94 commented:
Hello @dschult, I was able to compile the expected.tex on overleaf: https://www.overleaf.com/read/mpyqdcyypfgx

LucaCappelletti94 commented:
> I removed the conflicts with the master branch and reformatted docstrings and switched names from vertices/vertex to nodes/node. Also format changes to match PEP8, and removed numpy requirements for the tests, as well as making the test class into a test function. I also made some cosmetic code changes.
> 
> Questions/concerns/suggestions:
> 
> * Should we make a public function like `write_latex(G, path)` to ease use?

I guess it depends on the general expected user usage of the Networkx library. This class was meant to be a stand-alone conversion from a graph to Latex, in a way that it can be easily customized. I guess that a method such as the plot methods that are already present could be used, as no heavy lifting is executed in the constructor of the class.

> * There are other aspects of the public-facing API that raise questions... Does the basic code need to be inside a figure environment?  Should it use subfigures at all (that's for putting together many graphs... which presumably should be done by another program using this one to generate the latex code to draw just the one figure).

Yes, there is no strict need for figures nor subfigures, but I always found myself in need for one.

> * Why does the code cut the x and y layout sizes in half?

I honestly do not remember exactly the meaning of that: I seem to remember that it was needed to properly centre the graph.

> * Could we remove the placeholder files and replace them with text strings in nx_latex.py? The files are small.  Is there another functionality you aim to include by allowing placeholder files?

I don't think there are contro-indications to put the files into strings, they just felt more modular being separated this way.

> * Could the test compare the strings generated rather than writing them to files and comparing those?  I think  this means, including the expected.tex file as a string and comparing that string to the string generated by `str(A)`.

Yes that definitely makes sense.

> * Will this code work for DiGraph and MultiGraph?

It will work for DiGraphs but not for MultiGraphs.

> * Can we make some examples in the doc_string? Or in the examples gallery (directory examples/drawing)?

Sure, but I'd need some help to get started with these.

> 
> Thanks!

Thank you for your work and sorry for my late reply. This period has been filled to the brink of stuff to do.

Have a nice day!

dschult commented:
Thanks very much for this. It looks like my version of adigraph is 1.5.1 and the latest is 1.7.1.  That may be the cause of the xparse error complaining about extra commas. Does this make sense?

Is there a version of adigraph that we should require for this to work?

LucaCappelletti94 commented:
> Thanks very much for this. It looks like my version of adigraph is 1.5.1 and the latest is 1.7.1. That may be the cause of the xparse error complaining about extra commas. Does this make sense?
> 
> Is there a version of adigraph that we should require for this to work?

When I developed the python wrapper I was using 1.7.x, so I think we should use that one.

I would expect for any updated version of the LaTex distribution to have the latest one (as even overleaf has it, and they update very rarely), considering running the following to update your package:

```bash
sudo tlmgr update --self
sudo tlmgr update adigraph
```

On an unrelated note, I'm finalizing a python + rust backend graph library which has some features that might interest you if you'd like to have a chat after we finish up this pull request.

dschult commented:
This PR was closed since Feb apparently due to the original fork being deleted.
I reopened it by creating PR #5639 with these changes. 
The adigraph package is still available on CTAN (written by the author of this PR) so it should all still work.

LucaCappelletti94 commented:
Hello @dschult , I must have done that by mistake. Likely I thought it was a spurious fork. Do you need help putting it back together?

dschult commented:
No, it transferred to the new PR just fine.  Looks like it'll need updating/rebasing to get the CI to pass. but I think that will be straightforward.  Thanks!
